File: order-service/src/main/java/orderservice/order_service/entity/Discount.java
package orderservice.order_service.entity;

import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "discounts")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Discount {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "discount_id")
    Integer discountId;

    @Column(name = "code", nullable = false, unique = true, length = 50)
    String code;

    @Column(name = "name", nullable = false, length = 100)
    String name;

    @Column(name = "description", length = 255)
    String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "discount_type", nullable = false)
    DiscountType discountType;

    @Column(name = "discount_value", nullable = false, precision = 10, scale = 2)
    BigDecimal discountValue;

    @Column(name = "min_order_amount", precision = 12, scale = 2)
    BigDecimal minOrderAmount;

    @Column(name = "max_discount_amount", precision = 12, scale = 2)
    BigDecimal maxDiscountAmount;

    @Column(name = "start_date", nullable = false)
    LocalDateTime startDate;

    @Column(name = "end_date", nullable = false)
    LocalDateTime endDate;

    @Column(name = "usage_limit")
    Integer usageLimit;

    @Column(name = "used_count")
    Integer usedCount;

    @Column(name = "branch_id")
    Integer branchId;

    @Column(name = "active", nullable = false)
    Boolean active;

    @Column(name = "create_at", nullable = false)
    LocalDateTime createAt;

    @Column(name = "update_at", nullable = false)
    LocalDateTime updateAt;

    @PrePersist
    protected void onCreate() {
        createAt = LocalDateTime.now();
        updateAt = LocalDateTime.now();
        if (usedCount == null) {
            usedCount = 0;
        }
        if (active == null) {
            active = true;
        }
        if (minOrderAmount == null) {
            minOrderAmount = BigDecimal.ZERO;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updateAt = LocalDateTime.now();
    }

    public enum DiscountType {
        PERCENT, AMOUNT
    }
}

File: fe_coffee_manager/src/types/discount.ts
export interface Discount {
    discountId: number;
    code: string;
    name: string;
    description?: string;
    discountType: 'PERCENT' | 'AMOUNT';
    discountValue: number;
    minOrderAmount: number;
    maxDiscountAmount?: number;
    startDate: string;
    endDate: string;
    usageLimit: number;
    usedCount: number;
    branchId?: number;
    active: boolean;
    createAt: string;
    updateAt: string;
}

export interface CreateDiscountRequest {
    code: string;
    name: string;
    description?: string;
    discountType: 'PERCENT' | 'AMOUNT';
    discountValue: number;
    minOrderAmount?: number;
    maxDiscountAmount?: number;
    startDate: string;
    endDate: string;
    usageLimit?: number;
    branchId?: number | null;
    active?: boolean;
}

export interface UpdateDiscountRequest {
    name?: string;
    description?: string;
    discountType?: 'PERCENT' | 'AMOUNT';
    discountValue?: number;
    minOrderAmount?: number;
    maxDiscountAmount?: number;
    startDate?: string;
    endDate?: string;
    usageLimit?: number;
    branchId?: number | null;
    clearBranch?: boolean;
    active?: boolean;
}

export interface ApplyDiscountRequest {
    discountCode: string;
    orderAmount: number;
    branchId: number;
}

export interface DiscountApplicationResponse {
    discountCode: string;
    discountName: string;
    discountType: string;
    discountValue: number;
    originalAmount: number;
    discountAmount: number;
    finalAmount: number;
    isValid: boolean;
    message: string;
}

export interface DiscountPageResponse {
    content: Discount[];
    page: number;
    size: number;
    totalElements: number;
    totalPages: number;
    first: boolean;
    last: boolean;
}
